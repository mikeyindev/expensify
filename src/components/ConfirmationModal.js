import React from 'react';
import Modal from 'react-modal';

Modal.setAppElement(document.getElementById("app"));

const ConfirmationModal = (props) => {

  // console.log(props);
  return (
    // onRequestClose allows users to click outside the modal or hit the esc button to close the modal
    <Modal
      isOpen={props.modalIsOpen}
      onRequestClose={props.closeModal}
      // closeTimeoutMS={200}
      className="ConfirmationModal"
    >
      <h3 className="ConfirmationModal__title">
        Are you sure you want to delete the expense?
      </h3>
      <button
        className="button"
        id="ConfirmationModal__button"
        onClick={props.startRemoveExpense}
      >
        Yes
      </button>
      <button className="button--remove" onClick={props.closeModal}>
        No
      </button>
    </Modal>
  );
};

export default ConfirmationModal;