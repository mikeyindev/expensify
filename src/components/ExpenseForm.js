import React from 'react';
import moment from 'moment';
import 'react-dates/initialize';
import { history } from '../routers/AppRouter';
import { SingleDatePicker } from 'react-dates';

// const now = moment();
// console.log(now.format('MMM Do, YYYY'));

// SingleDatePicker props:
// numberOfMonths sets the number of months to show in the calendar at the same
// time.
// Setting isOutsideRange to 'false` makes every single date available.

// This component has local state, so we use a class component
export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '',
      // Convert back from number to string
      amount: props.expense ? (props.expense.amount / 100).toString() : '',
      // Convert back from milliseconds to a moment object
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: ''
    };
    console.log(history.location.pathname);
  }

  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };

  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };

  onAmountChange = (e) => {
    const amount = e.target.value;
    // Regex for limiting digits to optional 2 decimal places, with at least 1
    // digit before decimal.
    // `!amount` allows users to highlight the value and delete it, so it
    // matches and empty string
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };

  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };

  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };

  onSubmit = (e) => {
    e.preventDefault(); // prevent page refresh
    // Form validation
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({ error: 'Please provide description and amount' }));
    } else {
      this.setState(() => ({ error: '' }));

      this.props.onSubmit({
        description: this.state.description,
        // this.state.amount is string and needs to be converted to float
        amount: parseFloat(this.state.amount, 10) * 100,
        // this.state.createdAt is a moment object and needs to be converted to
        // milliseconds for the time stamp
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });

      console.log('submitted');
    }
  };

  render() {
    return (
      <form className="form" onSubmit={this.onSubmit}>
        {this.state.error && <p className="form__error">{this.state.error}</p>}
        <div className="input-group">
          <div className="input-group__item">
            <input
              className="input"
              type="text"
              value={this.state.description}
              onChange={this.onDescriptionChange}
              placeholder="Description"
              autoFocus
            />
          </div>
          <div className="input-group__item">
            <input
              className="input"
              type="text"
              value={this.state.amount}
              onChange={this.onAmountChange}
              placeholder="Amount"
            />
          </div>
          <div className="input-group__item">
            <SingleDatePicker
              date={this.state.createdAt}
              focused={this.state.calendarFocused}
              isOutsideRange={() => {
                false;
              }}
              numberOfMonths={1}
              onDateChange={this.onDateChange}
              onFocusChange={this.onFocusChange}
            />
          </div>
          <div className="input-group__item">
            <textarea
              className="textarea"
              onChange={this.onNoteChange}
              placeholder="Add a note for your expense (optional)"
              value={this.state.note}
            />
          </div>
        </div>
        {
          
          history.location.pathname.includes('/edit/') ? (
            <div className="button-group">
              <button className="button">Save Changes</button>
              <button
                className="button--remove"
                onClick={this.props.startRemoveExpense}
              >
                Remove Expense
              </button>
            
            </div>
          ) : (
            <div className="button-group">
              <button className="button">Add Expense</button>
            </div>
          )
        } 
      </form>
    );
  }
}
