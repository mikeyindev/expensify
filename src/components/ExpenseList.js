import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import filterExpenses from '../selectors/filterExpenses';

// Presentational component
const ExpenseList = (props) => (
  <div>
    <h1>Expense List</h1>
    {props.expenses.map((expense) => (<ExpenseListItem key={expense.id} {...expense} />))}
  </div>
);

// Convert state to props passed to ExpenseList
const mapStateToProps = (state) => {
  return {
    expenses: filterExpenses(state.expenses, state.filters)
  }
};

// connect() returns a function and we're passing the ExpenseList component to
// the returned function
export default connect(mapStateToProps)(ExpenseList);