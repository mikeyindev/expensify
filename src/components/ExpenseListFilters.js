import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filterActions';
import { DateRangePicker } from 'react-dates';

// Presentational component that allows us to set the text filter from the UI.
// The new text filter is written to the store
class ExpenseListFilters extends React.Component {
  state = {
    calendarFocused: null
  };

  onDatesChange = ({ startDate, endDate }) => {
    this.props.dispatch(setStartDate(startDate));
    this.props.dispatch(setEndDate(endDate));
  };

  onFocusChange = (calendarFocused) => {
    this.setState(() => ({ calendarFocused }))
  };

  render() {
    return (
      <div>
        <input 
          type="text" 
          value={this.props.filters.text} 
          onChange={(e) => {
            this.props.dispatch(setTextFilter(e.target.value));
            console.log(e.target.value)
          }} 
        />
        <select 
          value={this.props.filters.sortBy} 
          onChange={(e) => {
            if (e.target.value === 'date') { 
              this.props.dispatch(sortByDate()); 
            } 
            else if (e.target.value === 'amount') { 
              this.props.dispatch(sortByAmount()); 
            }
          }}
        >
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        <DateRangePicker 
          startDate={this.props.filters.startDate} 
          startDateId='start_date'
          endDate={this.props.filters.endDate}
          endDateId='end_date'
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
          showClearDates={true}
        />
      </div>
    );
  }
}

const mapStateToProps = (state) => ({ filters: state.filters });

export default connect(mapStateToProps)(ExpenseListFilters);