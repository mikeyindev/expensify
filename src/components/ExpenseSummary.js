import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import filterExpenses from '../selectors/filterExpenses';
import selectExpenseTotal from '../selectors/expenseTotal';

export const ExpenseSummary = ({ expenseCount, expenseTotal }) => {
  const expenseWord = expenseCount === 1 ? 'expense' : 'expenses';
  const formattedExpenseTotal = numeral(expenseTotal / 100).format('$0,0.00');
  return (
    <div>
      <h1>
        Viewing {expenseCount} {expenseWord} totalling {formattedExpenseTotal}
      </h1>
    </div>
  )
};

const mapStateToProps = (state) => {
  const filteredExpenses = filterExpenses(state.expenses, state.filters);
  return {
    expenseCount: filteredExpenses.length,
    expenseTotal: selectExpenseTotal(filteredExpenses)
  };
};

export default connect(mapStateToProps)(ExpenseSummary);