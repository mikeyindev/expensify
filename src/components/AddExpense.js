import React from 'react';
import { connect } from 'react-redux';
import { startAddExpense } from '../actions/expenseActions';
import ExpenseForm from './ExpenseForm';

export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    // console.log(expense);
    // The expense object is passed in by ExpenseForm when it calls onSubmit(). The expense object is the updated state.
    this.props.startAddExpense(expense);
    // Redirect after submitting form without page refresh
    this.props.history.push("/");
    console.log(expense);
  };

  render() {
    return <div className="page-header">
        <div className="content-container">
          <h1 className="page-header__title">Add Expense</h1>
          <ExpenseForm onSubmit={this.onSubmit} />
          <div className="button-group">
            <button className="button" form="expense-form" type="submit">
              Add Expense
            </button>
          </div>
        </div>
      </div>;
  }
}

const mapDispatchToProps = (dispatch) => ({
  startAddExpense: (expense) => dispatch(startAddExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);