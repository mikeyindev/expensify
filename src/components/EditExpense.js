import React from 'react';
import { connect } from 'react-redux';
import { editExpense, startRemoveExpense } from '../actions/expenseActions';
import ExpenseForm from './ExpenseForm';

export class EditExpensePage extends React.Component {
  // The expense object is passed in by ExpenseForm when it calls onSubmit().
  // The expense object contains the updated state but doesn't include id. That
  // why we still need to use props.expense.id instead of expense.id. 
  onSubmit = (expense) => {
    this.props.editExpense(this.props.expense.id, expense);
    this.props.history.push('/');
    // console.log('Updated', expense);
  };

  startRemoveExpense = () => {
    this.props.startRemoveExpense({ id: this.props.expense.id });
    this.props.history.push('/');
  }

  render() {
    return (
      <div>
        <ExpenseForm 
          expense={this.props.expense} 
          onSubmit={this.onSubmit} 
        />
        <button 
          onClick={this.startRemoveExpense}
        >Remove</button>
      </div>
    );
  }
}

// with connect(), we have access to both the state (in the Redux store) and the
// props passed to the component by the higher order component, Provider.
const mapStateToProps = (state, props) => {
  // Returning a new prop to the EditExpensePage component
  return {
    expense: state.expenses.find((expense) => 
      expense.id === props.match.params.id)
  };
};

const mapDispatchToProps = (dispatch, props) => ({
  editExpense: (id, expense) => dispatch(editExpense(id, expense)),
  startRemoveExpense: (id) => dispatch(startRemoveExpense(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);