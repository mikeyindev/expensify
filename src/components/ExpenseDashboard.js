import React from 'react';
import ExpenseList from "./ExpenseList";
import ExpenseListFilters from './ExpenseListFilters';
import ExpenseSummary from './ExpenseSummary';
import { handleHotkey } from '../util';

const ExpenseDashboardPage = () => {
  document.addEventListener('keydown', (e) => handleHotkey(e));
  return (
    <div>
      <ExpenseSummary />
      <ExpenseListFilters />
      <ExpenseList />
    </div>
  );
};

export default ExpenseDashboardPage;
