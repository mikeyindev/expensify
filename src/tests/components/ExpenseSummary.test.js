import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseSummary } from '../../components/ExpenseSummary';
import selectExpenseTotal from '../../selectors/expenseTotal';
import expenses from '../fixtures/expenses';
import expenseTotal from '../../selectors/expenseTotal';

test('should correctly render ExpenseSummary with 1 expense', () => {
  const wrapper = shallow(<ExpenseSummary 
    expenseCount={1} 
    expenseTotal={selectExpenseTotal([expenses[0]])} 
    />);
  expect(wrapper).toMatchSnapshot();
});

test('should correctly render ExpenseSummary with 1 expense', () => {
  const wrapper = shallow(<ExpenseSummary
    expenseCount={expenses.length}
    expenseTotal={selectExpenseTotal(expenses)}
  />);
  expect(wrapper).toMatchSnapshot();
});

