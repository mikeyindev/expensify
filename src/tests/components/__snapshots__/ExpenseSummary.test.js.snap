// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render ExpenseSummary with 1 expense 1`] = `
<div>
  <h1>
    Viewing 
    1
     
    expense
     totalling 
    $1.95
  </h1>
</div>
`;

exports[`should correctly render ExpenseSummary with 1 expense 2`] = `
<div>
  <h1>
    Viewing 
    3
     
    expenses
     totalling 
    $241.95
  </h1>
</div>
`;
